### ISSUE TITLE
add support for 'managed' dependencies?

### ISSUE THREAD

**Issue by cprice404** [technomancy/leiningen#2067]
In Maven, a parent pom can specify a section called `dependencyManagement`. If this is specified, the versions of the dependencies listed in this section are used when resolving transitive dependencies. It also allows you to specify dependencies in the main sections of the pom w/o specifying a version number for them.

Exposing this functionality through leiningen would be useful for writing leiningen plugins that did things analogous to what maven's "parent poms" can do.  I'm imagining that this would probably entail adding a new key called `:managed-dependencies` and then a little work to merge that in with the regular dependencies before calling into `pomegranate`/`aether` to do the actual dependency resolution.

(Note: `pomegranate` doesn't support this yet, but I've poked at the code and it seems like it would be very easy to add.  I captured that in https://github.com/cemerick/pomegranate/issues/72 .)

If a top-level `:managed-dependencies` section didn't seem palatable, I think there are some other ways this could be implemented by just refactoring the code that determines which keys from the project get passed in to the various functions that do the dependency resolution.  Then a lein plugin could add hooks to those functions to add in additional keys that it had defined its own support for.

At this point I'm just exploring this idea in an attempt to solve some problems we're having at our company about how to manage the version numbers of super-common dependencies that end up getting used by tons of our projects.  If there was consensus that making this possible in lein was worthwhile, I'm happy to take input on what the best implementation paths might be, and put together a PR if we actually end up pushing forward on this.

**Comment by hyPiRion** [comment]
It sounds like this is possible through middleware, or am I wrong here?

I think this sounds like reasonable functionality to have for these kinds of projects, but it seems like yet another concept Lein users may have to learn. So even though I think this seems useful, I'd like to avoid including it before a bigger audience would like this in. 

If it's not possible to do this through middleware, then it may be more reasonable to push it into Leiningen.

**Comment by cprice404** [comment]
I definitely hear you about wanting to minimize cognitive burden for users.  In the interest of that I'd definitely be happy enough just trying to figure out a way to be able to add support for this for a plugin.

Unfortunately, I don't think it's possible through middleware given the current lein code.  "Managed" dependencies are conceptually different than regular dependencies, in that they don't actually cause a dependency to be resolved and brought into the project; they only specify a version number to be used if something *else* brings the dependency into the project.

So, e.g., if it were possible to specify managed dependencies in lein and I had a list that included `[clj-time "0.10.0"]` and `[core.async "1.0"]`, but then the *regular* dependencies section only specified [clj-time], then we wouldn't actually end up dragging in the `core.async` dependency... just enforcing that the version of the clj-time dependency was correct.  I don't think this kind of effect can be achieved w/o actually hooking in to the `managedDependencies` features of maven/aether.

I think that the key is being able to pass in a value for the [`managedDependencies` argument of Aether's `CollectRequest`](http://download.eclipse.org/aether/aether-core/1.0.1/apidocs/org/eclipse/aether/collection/CollectRequest.html#CollectRequest%28java.util.List,%20java.util.List,%20java.util.List%29).

That would need to happen in the [`cemerick.pomegranate.aether/resolve-dependencies*` function](https://github.com/cemerick/pomegranate/blob/pomegranate-0.3.0/src/main/clojure/cemerick/pomegranate/aether.clj#L710-L717), where it is currently hard-coded to `nil`.  I have a branch where I've got a working proof-of-concept of adding support for managed dependencies in pomegranate.

From there it would be a matter of figuring out how to add some data to [`lein`'s calls into `pomegranate`, from ](https://github.com/technomancy/leiningen/blob/d2a2076477e0a620cf963171daef1d2d136a957a/leiningen-core/src/leiningen/core/classpath.clj#L115-L117).

Without some tweaks to that `leiningen.core.classpath/get-dependencies-memoized` fn, the argument list that is passed in to the `pomegranate` functions is hard-coded and can't include any keys from the project [except the ones that are hard-coded into the function](https://github.com/technomancy/leiningen/blob/d2a2076477e0a620cf963171daef1d2d136a957a/leiningen-core/src/leiningen/core/classpath.clj#L110-L111).

I can imagine refactoring that namespace to create a new function called, e.g., `get-additional-pomegranate-opts`, and then having lein call that and merge the result in with its hard-coded list of pomegranate opts before actually calling into pomegranate.  Then, a user such as myself could write a plugin that added a hook around `get-additional-pomegranate-opts` to add in the additional stuff that I need, and I think this would end up being pretty much transparent to the end user.  I'd be happy to try to scratch out a POC PR that goes this direction if it sounded tolerable.

**Comment by cprice404** [comment]
One thing that could be a nice benefit of adding support for managed dependencies directly into `lein`, as opposed to in a plugin is that projects could opt to leave out the `version` field in their dependencies list altogether if the were using managed dependencies (presumably through some kind of parent project description).

So, the idea is that you'd have a parent project that had something like this:

```clj
(defproject my-parent-project "1.0.0"
  :description "Parent project"
  :managed-dependencies [[clj-time "0.10.0"]
                                            [core.async "1.0"]])
```

Then you could have child projects `foo` and `bar`:

```clj
(defproject foo "1.0.0"
   :parent-project [my-parent-project "1.0.0"]
   :dependencies [[clj-time]
                            [some-other-dep "2.0.0"]])
(defproject bar "1.0.0"
   :parent-project [my-parent-project "1.0.0"]
   :dependencies [[core.async]
                            [yet-another-dep "3.0.0"]])
```

In this example, `foo` would automatically get the correct version of `clj-time` from the parent, but would not drag in `core.async` at all.  Likewise, `bar` would get the correct version of `core.async` but would not drag in `clj-time`.  When you wanted to bump to a newer version of `clj-time` or `core.async` across all of your projects, you'd just update the parent project.

Not having to specify the version number of the dependencies in the child projects seems like a pretty big win for setups where you have a large number of projects / repos / libraries that make up your application.  That's definitely happening to us at Puppet Labs lately, and it seems like it's happening elsewhere as well (e.g. https://github.com/walmartlabs/shared-deps ).  It seems like hooking into maven/aether's pre-existing solution for this problem might be a more complete and maintainable approach than trying to re-invent the same things w/o leveraging `managedDependencies`.

From my initial poking it seems to me that the existing lein code is already written in a way that is pretty tolerant of omitting the `version` field from a dependency tuple.

Anyway... just trying to provide additional context here.  I recognize that this probably seems like a pretty big change I'm hinting at, and I totally understand the hesitance w/rt adding more cognitive burden for users.  I do think that this stuff could be added in a pretty transparent way that users didn't have to worry about unless they needed it, w/o breaking any backward compatibility.  I also can imagine a path forward on this stuff where it starts off as a plugin and then is considered for migration into the core in the future if it's working well.  I do think the patches to `pomegranate` and  `lein.core.classpath/get-dependencies-memoized` would be necessary before a plugin would even work, though.

Alright, I'll shut up now :)  Thanks for your consideration!

**Comment by hyPiRion** [comment]
Ah, with an example and the fact that it's seemingly hard/impossible to do this with middleware, this seems far more reasonable to add in. Happy to take a patch that implements it in the way shown in the example you provide. :)

I see that parent poms may contain the entry `relativePath` in Maven. Is this something you'd like to add in as well?

**Comment by cprice404** [comment]
Ah, so you'd be willing to take a patch that added `managed-dependencies` and `parent-project` into core, as opposed to via a plugin?

Wasn't necessarily planning on supporting something like `relativePath` out of the gate, but maybe that'd be worthwhile?  I know that the existing `lein-parent` plugin, which does a subset of this kind of thing, *requires* that you use a relative path, so maybe that is something that people would want.

In any case, it sounds like the next step is to get a patch in for `pomegranate`.  Thanks again for your feedback.

**Comment by hyPiRion** [comment]
Yeah, it seems a bit clearer now that you have explained it, the rationale and shown an example. Since it's an argument to the dependency resolver, I think it makes sense to add into Leiningen proper instead of making a plugin that hooks into the dependency resolver (bound to cause issues).

I don't think `relativePath` is a necessary thing to include right off the bat. I think I'd prefer to start without it and see if there's a real need for it.

**Comment by cprice404** [comment]
Now that the pomegranate patch is in, this is unblocked and I can probably start working on it in the next week or so if that still sounds OK to you, @hyPiRion .

If so, I would probably approach this as two separate PRs.  The first one would simply add support for a `managed-dependencies` section in the `project.clj` directly, without tackling any of the `parent-project` stuff.  The second one would add in some initial support for `parent-project`.

**Comment by cprice404** [comment]
Linking in initial PR just for posterity: #2126

**Comment by cprice404** [comment]
@hyPiRion I know you're busy, and I'm still working on getting help w/review from my co-workers.  So no sweat if you don't have time to respond to this for a while... but... I started looking at implementing the "parent" project portion of the features described here, and, if we merged my current PR that adds support for the "managed-dependencies" section, it looks to me like the "parent project" part could perhaps be handled entirely in middleware.  In fact, it looks like this existing plugin already does the majority of the rest of what I'd want to do:

https://github.com/achin/lein-parent/blob/master/src/leiningen/parent.clj

The only obvious thing that it's missing in my mind is the ability to resolve the parent project as a maven artifact, rather than requiring a relative path.

So, I think I may proceed by forking that plugin and hacking on the maven resolution piece of it, and if it all works, it might make sense to just stick with a plugin for that part for now?  And re-consider moving it into the core later?  Would be curious to hear your thoughts.

**Comment by hyPiRion** [comment]
@cprice404: I would prefer to eventually get it into core, as managed-dependencies/parent-project is part of dependency management. However, we're in no rush of doing that, and if it's easier to make it as a middleware-plugin then we can start out there.

**Comment by cprice404** [comment]
@hyPiRion cool.  I'm thinking that maybe doing it as a plugin for now would be smart since it would allow us to tinker with APIs and stuff before committing to anything permanent in the core.  Of course, for that to be a tractable path, we'd need to be willing to do a lein release that contains the managed_dependencies part on its own first.

I have a branch of the existing lein-parent plugin now that seems to work and prove out the use case for managed_dependencies, but it relies on #2164 and #2126 being in a released version of leiningen first.

**Comment by cprice404** [comment]
Also, some day if we cross paths at a clojure conference or something, I would like to buy you a beer or three :)

**Comment by glts** [comment]
A pull request addressing this was merged in #2126. Managed dependencies are now supported according to the sample.project.clj and https://github.com/technomancy/leiningen/blob/master/doc/MANAGED_DEPS.md.

With that I think we can close this issue, and/or open new ones if there is anything not yet resolved here.
